src/re/multi_grid/kernel.py

```diff
@@ -0,0 +1,312 @@
+                ker = jnp.concatenate(tuple(kk.ravel() for kk in ker))
+                return ker
+
+            @jit
+            def scanned_amend_unique(carry, idx, shift):
+                (u, inv) = carry
+                k = get_matrices(idx)
+                u, invid = amend_unique_(u, k, axis=0, atol=atol, rtol=rtol)
+                inv = inv.at[idx - shift].set(invid)
+                return (u, inv), invid
+
+            indices = grid_at.refined_indices()
+            indices = gridf_at.index2flatindex(indices)[0].ravel()
+            shift = np.min(indices)
+            size = np.max(indices) - shift
+            inv = np.full(size, buffer_size + 1)
+
+            shp = eval_shape(get_matrices, indices[0]).shape
+            unique = jnp.full((buffer_size,) + shp, jnp.nan)
+
+            (unique, inv), invid = scan(
+                partial(scanned_amend_unique, shift=shift), (unique, inv), indices
+            )
+            _, idx = np.unique(invid, return_index=True)
+            n = idx.size
+            if n >= unique.shape[0] or not np.all(np.isnan(unique[n:])):
+                raise ValueError("`mat_buffer_size` too small")
+            uids = indices[idx]
+            uids = gridf_at.flatindex2index(uids[np.newaxis, :])
+            uindices.append(uids)
+            invindices.append(inv)
+            indexmaps.append(_IdxMap(shift, gridf_at.index2flatindex))
+
+        return self.replace(
+            _cim=_CompressedIndexMap(
+                base_kernel=None,
+                kernels=None,
+                uindices=uindices,
+                indexmaps=indexmaps,
+                invindices=invindices,
+            )
+        )
+
+    def compress_matrices(self):
+        """Recompute the kernel matrices while keeping the tables for speedy lookups
+        fixed.
+
+        This is useful for updating the covariance function of a kernel without
+        recreating the tables in an expensive optimization.
+        """
+        assert self._cim is not None
+
+        base_kernel = self.compute_matrices(jnp.array([-1]), None)
+        kernels = tuple(
+            self.compute_matrices(ii, ll) for ll, ii in enumerate(self._cim.uindices)
+        )
+        cim = self._cim._replace(base_kernel=base_kernel, kernels=kernels)
+        return self.replace(_cim=cim)
+
+    def compress(self, *args, **kwargs):
+        """Optimize the kernel matrices for the given grid.
+
+        This links kernel matrices by similarity and subsequently looks them up via
+        tables. Compressing the kernel in this way has the potential to drastically
+        reduce the memory requirements of applying it and make the inference of
+        kernels faster.
+        """
+        return self.compress_indices(*args, **kwargs).compress_matrices()
+
+
+def _suitable_window_size(grid_at_level, default=3) -> Tuple[int]:
+    wsz = []
+    for g in grid_at_level.raw_grids:
+        if isinstance(g, HEALPixGridAtLevel):  # special-case HEALPix
```

Comment:
Happy to leave this as is, but maybe refer to as `default_window_size` instead? `HEALPixGrid` also works with a window size of 1 for example (which is definitely an edge case)